_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,203 @@
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/job-graph.png" width="800px" alt="Figure 2: Job Graph of the Fraud Detection Flink Job"/>
+<br/>
+<i><small>Figure 2: Job Graph of the Fraud Detection Flink Job</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/job-graph.png)
+
+The main blocks of the Transactions processing pipeline are:<br>
+
+* **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>
+
+* **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` operator hashes this dynamic key and partitions the data accordingly among all parallel instances of the following task
+
+* **Dynamic Alert Function** that accumulates a data window and creates Alerts based on it
+
+The job graph also depicts various data exchange patterns between the operators.
+
+* The __FORWARD__ connection following the Transaction Source means that all data consumed by one of the parallel instances of the Transaction Source operator is transferred to exactly one instance of the subsequent `DynamicKeyFunction` operator, without redistribution. It also indicates the same level of parallelism of the two connected operators (12 in this case). This communication pattern is depicted in Figure 3. Orange circles represent transactions and dotted rectangles depict parallel instances of the operators (also called [*tasks*](https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#operator)).  
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/forward.png" width="800px" alt="Figure 3: Message passing across tasks: FORWARD"/>
+<br/>
+<i><small>Figure 3: Message passing across tasks: FORWARD</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/forward.png)
+
+* The __HASH__ connection between the `DynamicKeyFunction` and the `DynamicAlertFunction` means that for each message a hash code is calculated and messages are evenly distributed among available parallel instances of the following operator. Such connection needs to be explicitly "requested" from Flink by using a `keyBy` operator.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across tasks: HASH (via `keyBy`)"/>
+<br/>
+<i><small>Figure 4: Message passing across tasks: HASH (keyBy)</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/hash.png)
+
+* A __REBALANCE__ distribution is either caused by an explicit call to `rebalance()` or by a change of parallelism (12 -> 1 in case of the job graph from Figure 2). Calling `rebalance()` causes data to be repartitioned randomly and can help to mitigate data skew in certain scenarios.
```

Comment:
I guess .shuffle() also cause a REBALANCE, so probably good to mentioned it to, right?

Reply:
Actually, `keyBy()`, `partitionCustom()`, `rebalance()`, `rescale()`, and `shuffle()` are all different partitioning schemes which cause a network shuffle. Not sure which one(s) you want to describe here, but `rebalance()`, `rescale()`, and `shuffle()` are certainly in a group of non-key-based shuffling mechanisms.
* However, you may not want to put them under a same label or if so, then it shouldn't be called "rebalance" because this is already a specific scheme our of these three.
* Be sure to describe the right one (the current description is wrong).
* Maybe also start with the "what" before you describe the "how"?

Reply:
test comment: REBALANCE is being used when changing the parallelism (one may think, this leads to RESCALE)